### PR TITLE
STORM-3407 Storm 2.x ConstraintSolverStrategy backward compatibility

### PR DESCRIPTION
**Pull request by dandsager1**
For backward compatibility, support topology.ras.constraint.max.state.traversal for old topologies.

### CODE REVIEW

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java [hunk 1]
```diff
@@ -248,10 +247,22 @@ public SchedulingResult schedule(Cluster cluster, TopologyDetails td) {
         nodes = RAS_Nodes.getAllNodesFrom(cluster);
         Map<WorkerSlot, Set<String>> workerCompAssignment = new HashMap<>();
         Map<RAS_Node, Set<String>> nodeCompAssignment = new HashMap<>();
-        //set max number of states to search
-        final int maxStateSearch = Math.min(MAX_STATE_SEARCH,
-                                            ObjectReader.getInt(td.getConf().get(Config.TOPOLOGY_RAS_CONSTRAINT_MAX_STATE_SEARCH),
-                                                                DEFAULT_STATE_SEARCH));
```
Comment (kishorvpatil) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java]: We still need defaults.

Comment (govind-menon) [reply storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java]: The default has been moved to defaults.yaml as a DaemonConfig

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java [hunk 2]
```diff
@@ -248,10 +247,22 @@ public SchedulingResult schedule(Cluster cluster, TopologyDetails td) {
...
+
+        //set max number of states to search maintaining backward compatibility for old topologies
+        String stormVersionString = td.getTopology().get_storm_version();
+        boolean isNewTopology = stormVersionString != null && stormVersionString.startsWith("2");
+
+        Object confMaxStateSearch = null;
+        if (isNewTopology == false) {
```
Comment (govind-menon) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java]: nit: rename to is2xTopology

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java [hunk 3]
```diff
@@ -290,15 +301,15 @@ public SchedulingResult schedule(Cluster cluster, TopologyDetails td) {
         }
 
         //early detection/early fail
-        if (!checkSchedulingFeasibility()) {
+        if (!checkSchedulingFeasibility(daemonMaxStateSearch)) {
```
Comment (govind-menon) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java]: nit: rename it to maxStateSearch (it's just the maxStateSearch that has been determined from the config for this topology)

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java (hunk 1 above)
Comment (dandsager1) [reply storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java]: The DEFAULT_STATE_SEARCH was added as Config.TOPOLOGY_RAS_CONSTRAINT_MAX_STATE_SEARCH; The larger MAX_STATE_SEARCH was added as DaemonConfig.RESOURCE_AWARE_SCHEDULER_MAX_STATE_SEARCH.

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java (hunk 2 above)
Comment (dandsager1) [reply storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java]: ok

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java (hunk 3 above)
Comment (dandsager1) [reply storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java]: making this change. this is actually a bug fix as the previous code just used MAX_STATE_SEARCH

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java (hunk 1 above)
Comment (kishorvpatil) [reply storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java]: cool

**Pull request closed without merging**